src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs

```diff
@@ -65,6 +65,78 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
         return TryGetAnnotationsOfType(resource, out environmentVariables);
     }
 
+    /// <summary>
+    /// Get the environment variables from the given resource.
+    /// </summary>
+    /// <remarks>
+    /// This method is useful when you want to make sure the environment variables are added properly to resources, mostly in test situations.
+    /// This method has asynchronous behavior when <paramref name = "applicationOperation" /> is <see cref="DistributedApplicationOperation.Run"/>
+    /// and environment variables were provided from <see cref="IValueProvider"/> otherwise it will be synchronous.
+    /// </remarks>
+    /// <param name="resource">The resource to get the environment variables from.</param>
+    /// <param name="applicationOperation">The context in which the AppHost is being executed.</param>
+    /// <returns>The environment variables retrieved from the resource.</returns>
+    /// <example>
+    /// <code>
+    /// using var builder = await DistributedApplicationTestingBuilder.CreateAsync&lt;SampleShop_AppHost&gt;();
+    ///
+    /// builder.AddMongoDB("mongo")
+    ///     .WithEndpoint("tcp", e => e.AllocatedEndpoint = new AllocatedEndpoint(e, "localhost", 3000, containerHost))
+    ///     .WithMongoExpress();
+    ///
+    /// var env = await mongoExpress.GetEnvironmentVariableValuesAsync();
+    ///
+    /// Assert.Collection(env,
+    ///     e =>
+    ///         {
+    ///             Assert.Equal("ME_CONFIG_MONGODB_URL", e.Key);
+    ///             Assert.Equal($"mongodb://{containerHost}:3000/?directConnection=true", e.Value);
+    ///         },
+    ///         e =>
+    ///         {
+    ///             Assert.Equal("ME_CONFIG_BASICAUTH", e.Key);
+    ///             Assert.Equal("false", e.Value);
+    ///         });
+    /// </code>
+    /// </example>
+
+    public static async ValueTask<IReadOnlyDictionary<string, string>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource,
```

Comment:
Does this need to return an IReadOnlyDictionary? Can it just be a Dictionary?

Since this is a new object every time the method is called, and no one else references it, what is the point of returning a read-only object? If the caller wants to modify/add/remove from the dictionary, they will need to make a copy of it. This seems more like a `.ToDictionary()` method, which returns a copy.